src/libstd/sys/sgx/condvar.rs

```diff
@@ -1,42 +0,0 @@
-use crate::sys::mutex::Mutex;
-use crate::time::Duration;
-
-use super::waitqueue::{WaitVariable, WaitQueue, SpinMutex};
-
-pub struct Condvar {
-    inner: SpinMutex<WaitVariable<()>>,
-}
-
-impl Condvar {
-    pub const fn new() -> Condvar {
-        Condvar { inner: SpinMutex::new(WaitVariable::new(())) }
-    }
-
-    #[inline]
-    pub unsafe fn init(&mut self) {}
-
-    #[inline]
-    pub unsafe fn notify_one(&self) {
-        let _ = WaitQueue::notify_one(self.inner.lock());
-    }
-
-    #[inline]
-    pub unsafe fn notify_all(&self) {
-        let _ = WaitQueue::notify_all(self.inner.lock());
-    }
-
-    pub unsafe fn wait(&self, mutex: &Mutex) {
-        let guard = self.inner.lock();
-        mutex.unlock();
-        WaitQueue::wait(guard);
-        mutex.lock()
-    }
-
-    pub unsafe fn wait_timeout(&self, mutex: &Mutex, _dur: Duration) -> bool {
-        mutex.unlock(); // don't hold the lock while panicking
-        panic!("timeout not supported in SGX");
```

Comment:
@jethrogb Can we take the [SGX deadline/timeout issue](https://github.com/rust-lang/rust/pull/56410#pullrequestreview-221302094) in this thread? The PR becomes much more organized if we keep discussions on a given topic in a thread. It does not fill up the main PR thread, and it can be marked as resolved once done.

Yes, `parking_lot` is based on deadlines. They provide more accurate and reliable timeouts on most systems. Has less risk of drifting from spurious wakeups etc. Since this is core to the design of `parking_lot` it's not really feasible to make certain platforms use another method.

Since the current implementation already panics on `wait_timeout` I'm not sure how the PR as is would make matters worse than they already are? Users would still be fine if they don't use wait with timeouts, and they would still panic if they do.

Given that SGX neither has absolute time, nor timeout support, I'm not sure how we could do anything better than what we already do. How would you propose we solve the problem?

Reply:
> Since the current implementation already panics on wait_timeout I'm not sure how the PR as is would make matters worse than they already are?

The current API in `std` uses timeouts and gives the SGX target complete leeway in providing an implementation. This PR makes that worse by forcing the deadline approach on all targets, thereby removing the migration path towards a future implementation with timeouts.

Reply:
SGX implements `Instant` like `pub struct Instant(Duration);`. So in reality they are just durations. And the code proposed here for `Condvar::wait_timeout` calls `parking_lot::Condvar::wait_for` which in turn converts the given timeout to a deadline via `Instant::now().checked_add(timeout)`. We will end up with a deadline `Instant` that is just a `Duration` that is `timeout` further into the future than what `Instant::now()` returned. If we were to expose some way for the SGX `ThreadParker` to extract the `Duration` from the deadline `Instant`, is there no way it could do anything useful with it? As long as we know (roughly) what the first `Instant::now()` returned inside `wait_for` we could reverse the computation and end up with the original timeout again. And then implement a timeout based park from there.

Reply:
Yeah, after some more discussion with @jseyfried I think I'm now convinced that something based on `Instant` is fine, because we can always control how/when the user obtains the `Instant`. In future versions of SGX you can actually specify `Instant`s accurately internally. The weird/problematic case that I was intuitively trying to avoid is actually when using `Duration`.